Add admin manage article feature

mistydqwq opened the pull request:
# [Feature] Add Admin Panel Article Management

## **Background**
This PR adds a new **Article Management** feature to the admin panel, allowing administrators to search for articles in the database by URL and modify their fields.

## **Changes Introduced**
### **Frontend Changes**
- Added a new **`article/management/`** folder under `recnet/src/app/admin/`.
  - Created `article/management/page.tsx` for the main page.
  - Created `article/management/ArticleManagementForm.tsx` for the article editing form.
- Updated `AdminPanelNav` to include the new **Article Management** section.
- Modified `src/serve/routers/article.ts`:
  - Modified the `getArticleByLink` logic for fetching articles based on URL and using digital library or not .

### **Backend Changes**
- Added a new API: **`adminUpdateArticle`** to update article fields in the database.
- Modified `getArticleByLink` API:
  - Introduced a new request parameter:  
    ```ts
    boolean useDigitalLibrary
    ```
  - **Purpose**: Determines whether to fetch articles using the `digitalLibrary` service.
  - **Default Value**: `true` (ensuring no impact on existing behavior).
  - **If set to `false`**: The API retrieves articles directly from the database without using `digitalLibrary`.
- Updated related backend components:
  - **DTOs**  
  - **Repository layer**  
  - **Service layer**  
  - **`libs/recnet-api-model`**


## **How to Test**
1. Navigate to the **Admin Panel**.
2. Go to the newly added **Article Management** section.
3. Enter a **URL** to search for an article in the database.
4. Modify the necessary fields and submit the changes.
5. Verify that:
   - The article updates correctly in the database.
   - The new `useDigitalLibrary` parameter works as expected (retrieves articles correctly when set to `false`).

swh00tw commented:
Good to see your 1st PR. I will review it. By the way, can you make sure the Github action successfully passed? It seems like you have linting issue. You can conigure your IDE so that you can apply linter on saving the files.

By the way, when you run `git commit -m "..."`, the commit hook should automatically lint the staging files for you. How did you do the git commit?

swh00tw commented:
For quick fix, you can do 
```
nx lint <PROJECT_NAME> --fix
```
for `recnet`, `recnet-api`, and `recnet-api-model`.

And push commit.

joannechen1223 commented:
@mistydqwq Please remove the unused code and comments to keep the code tidy and ready for review. Thank you!

swh00tw commented:
I haven't started code review yet, just run on my machine. The UI looks great!!
There some some small nitpicks

1. For the "month" field, although we store 0-indexed number in DB. It's better to display 1-indexed number in UI (you can apply the offset behind the scene) for the sake of UX. (If you can use the same Dropdown component in `NewArticleForm.tsx` that would be even better)
<img width="754" alt="Screenshot 2025-02-04 at 5 12 55 PM" src="https://github.com/user-attachments/assets/fa848655-1bdd-4d40-bd51-fa2d9800d786" />
<img width="280" alt="Screenshot 2025-02-04 at 5 16 15 PM" src="https://github.com/user-attachments/assets/fcc5fbbc-8663-4663-bfe1-56490d02e447" />

2. After I click "Save changes", the form disappears. Is it intended behavior? I felt like the disappearing behavior is weird and leaving the form there should be better. @joannechen1223 any thoughts?

swh00tw commented on libs/recnet-api-model/src/lib/api/article.ts:
```diff
@@ -5,10 +5,29 @@ import { articleSchema } from "../model";
+
+
+// PATCH /articles/admin
+export const AdminUpdateArticleDtoSchema = z.object({
```
We have a certain naming convention for API schema. For example, this API is `PATCH /articles/admin`, we will name this schema `patchArticlesAdminRequestSchema`.

swh00tw commented on libs/recnet-api-model/src/lib/api/article.ts:
```diff
@@ -5,10 +5,29 @@ import { articleSchema } from "../model";
+  isVerified: z.boolean().optional(),
+});
+
+export type UpdateArticleDto = z.infer<typeof AdminUpdateArticleDtoSchema>;
```
We have a certain naming convention for API schema. For example, this API is PATCH /articles/admin, we will name this schema `PatchArticlesAdminRequest`. Please take a look at other API schema.

swh00tw commented on apps/recnet/src/server/routers/article.ts:
```diff
@@ -12,13 +13,35 @@ export const articleRouter = router({
+  //     });
+  //     return getArticlesResponseSchema.parse(data);
+  //   }),
+  updateArticleByLink: checkIsAdminProcedure
+    .input(AdminUpdateArticleDtoSchema)
+    .mutation(async (opts) => {
+      const { link, ...data } = opts.input;
+      const { recnetApi } = opts.ctx;
+
+      await recnetApi.patch(`/articles/admin?link=${link}`, data);
+    }),
```
Your current code doesn't allow the admin to modify the "link" field. To allow, you should change your code like this.
```suggestion
  updateArticleByLink: checkIsAdminProcedure
    .input(AdminUpdateArticleDtoSchema)
    .mutation(async (opts) => {
      const { link } = opts.input;
      const { recnetApi } = opts.ctx;

      await recnetApi.patch(`/articles/admin?link=${link}`, opts.input);
    }),
```

swh00tw commented on apps/recnet/src/app/admin/article/management/page.tsx:
```diff
@@ -0,0 +1,81 @@
+
+export default function ArticleManagementPage() {
+  const [searchUrl, setSearchUrl] = useState("");
+  const [foundArticle, setFoundArticle] = useState<any>(null);
```
```suggestion
  const [foundArticle, setFoundArticle] = useState<Article | null>(null);

```

You should use the type from `@recnet/recnet-api-model`. You can add a line on top of the file `import { Article } from "@recnet/recnet-api-model"`

swh00tw commented on apps/recnet/src/app/admin/article/management/ArticleManagementForm.tsx:
```diff
@@ -0,0 +1,209 @@
+import { trpc } from "@recnet/recnet-web/app/_trpc/client";
+import { AdminUpdateArticleDtoSchema } from "@recnet/recnet-api-model";
+
+type ArticleFormData = z.infer<typeof AdminUpdateArticleDtoSchema>;
```
You already have same type in `recnet-api-model/src/lib/api/article`.

swh00tw commented on apps/recnet/src/app/admin/article/management/ArticleManagementForm.tsx:
```diff
@@ -0,0 +1,209 @@
+                                        onSubmitFinish,
+                                        onCancel,
+                                      }: {
+  currentArticle: Partial<ArticleFormData>;
```
I think the type of `currentArticle` should be `Article | null`.

swh00tw commented on apps/recnet/src/app/admin/article/management/ArticleManagementForm.tsx:
There is a small bug for all the string fields in the form. For example, if the article's `doi` field is current `null` in database. When I type something in form's `doi` field and then delete them, and then click "Save changes", the `doi` field in DB for this article will be changed to `""` (empty string).

To fix this, you can check out how I handle this kind of edge case in `onboard/page.tsx`. You can essentially use `setValueAs` field to fix this. For example:
```tsx
<TextField.Root
    className="w-full"
    size="3"
    placeholder="(Optional)"
    {...register("url", {
      setValueAs: (value) => {
        if (value === "") {
          // when the input is empty string, set to null to avoid url validation
          return null;
        }
        return value;
      },
    })}
  >
```

swh00tw commented on apps/recnet-api/src/modules/article/dto/update.article.admin.dto.ts:
```diff
@@ -0,0 +1,51 @@
+    example: 7,
+    description: 'The publication month of the article (1-12).',
+  })
+  month?: number;
```
```suggestion
  month?: number | null;

```

swh00tw commented on apps/recnet-api/src/modules/article/dto/update.article.admin.dto.ts:
```diff
@@ -0,0 +1,51 @@
+    example: 'This article explores how AI can be used to improve patient outcomes...',
+    description: 'A brief abstract or summary of the article.',
+  })
+  abstract?: string;
```
```suggestion
  abstract?: string | null;

```

swh00tw commented on apps/recnet-api/src/modules/article/article.service.ts:
```diff
@@ -19,39 +20,68 @@ export class ArticleService {
   public async getArticleByLink(
-    link: string
+    link: string,
+    fetchFromDigitalLibrary: boolean = true
```
```suggestion
    useDigitalLibraryFallback: boolean = true
```
I think the naming can be better. It's confusing to name like this since if you set `fetchFromDigitalLibrary` to true, it will first fetch from our DB and if not found, it will fetch from Digital Library. Therefore, I think naming `fetchFromDigitalLibrary` is not accurate.

Just provide one of my random idea, we can name it `useDigitalLibraryFallback`. @joannechen1223 any thought?

mistydqwq replied:
Actually like what I commented out, at first I implemented a separated API to get article directly from the database, but our professor suggests me to combine the getDbArticleByLink with the original getArticleByLink because these two function are pretty similar and he doesn't want the API to explode at the beginning. Also make sense. So I changed the getArticleByLink function to the current version. May need to discuss.

swh00tw replied:
I think using separate method might be more scalable in this case given the original method was already really complicated. It also makes sense to try not to make too many APIs but I would prefer make functions easy to maintain than try to add complex logic in a function which is already heavy.
@joannechen1223 thoughts?

swh00tw replied:
@yoavartzi Yeah that's the small tradeoff. But I am worried about the over complicated logic inside single function, that's also not easy for future maintainers to understand. IMO, the small amount of duplication is okay.

yoavartzi replied:
Is there an option to create two APIs but with more code re-use? So the logic is not duplicated. You could think of the two APIs calling the same function that actually implements the logic (and that function itself not being exposed through an API)

swh00tw replied:
Sometimes there's this option but sometimes no. I think in this case, it is not really meaningful to do this since, if we are gonna create another API, the function body would look like this, it's already pretty small and not much thing you can abstract away 🤣 . I am actually good with current implementation but please keep/add the comments.

```ts
  let article = await this.articleRepository.findArticleByLink(link);
  if (article) {
    return { article };
  }
  return { article: null };
 ```

swh00tw reviewed (changes requested):
GJ overall. Just need to take care of some details. Btw, you can also add other teammates as reviewers!

swh00tw commented on apps/recnet-api/src/modules/article/article.service.ts:
```diff
@@ -19,39 +20,68 @@ export class ArticleService {
+  // }
+
   public async getArticleByLink(
-    link: string
+    link: string,
+    fetchFromDigitalLibrary: boolean = true
   ): Promise<GetArticleByLinkResponse> {
     let unifiedLink = link;
 
-    // If digital library service is available, try to get the unified link
-    if (this.digitalLibraryService) {
+    if (fetchFromDigitalLibrary && this.digitalLibraryService) {
       unifiedLink = await this.digitalLibraryService.getUnifiedLink(link);
     }
 
-    // Try to find the article in the database using the unified link
     let article = await this.articleRepository.findArticleByLink(unifiedLink);
     if (article) {
       return { article };
     }
 
-    // If no article found, try to get metadata using the digital library service
-    if (this.digitalLibraryService) {
+    if (fetchFromDigitalLibrary && this.digitalLibraryService) {
       try {
         const metadata = await this.digitalLibraryService.getMetadata(link);
         const articleInput = this.transformMetadata(metadata, unifiedLink);
-
         article = await this.articleRepository.createArticle(articleInput);
         return { article };
       } catch (error) {
         this.logger.error(error);
       }
     }
 
-    // If no article found and no metadata available, return null
-    return {
-      article: null,
-    };
+    return { article: null };
```
Btw, I think leaving the comment there would be helpful for developers to understand. Since the original logic might not be easy to understand.

joannechen1223 commented on apps/recnet-api/src/modules/article/article.controller.ts:
```diff
@@ -35,7 +37,57 @@ export class ArticleController {
+  @Auth({ allowedRoles: ["ADMIN"] })
+  @UsePipes(new ZodValidationBodyPipe(AdminUpdateArticleDtoSchema))
+  public async updateArticleByLink(
+    @Query("link") link: string,
```
It will be better to update the article by its unique **id** instead of the link, or some weird edge cases will happen. For example:
1. I put `A` URL in the searchbar and got the result. I changed `A` to `B`, while there was an existing row with link=`B` in our database. The result would become that the row with `A` link remained the same, and the row with `B` link was overwritten by the provided request data.
2. The statement in the UI says users can change the article URL, but it encountered not found error when I tried it.

joannechen1223 commented on apps/recnet-api/src/modules/article/article.controller.ts:
```diff
@@ -35,7 +37,57 @@ export class ArticleController {
+    @Query("link") link: string,
+    @Body() dto: AdminUpdateArticleDto
+  ) {
+    if (!link) {
+      throw new NotFoundException("Must provide ?link=xxx in query");
+    }
+
+    const existingArticle = await this.articleService.getArticleByLink(link, false);
+    if (!existingArticle) {
+      throw new NotFoundException(`Article not found by link=${link}`);
+    }
+
+    const updatedArticle = await this.articleService.updateArticleByLink(
+      link,
+      dto
+    );
+
+    return { article: updatedArticle };
```
In our code architecture design, we tried to encapsulate all the simple validation logic with Zod validation pipe and keep all the business logic(including more complex validation, throwing error, calling API, transforming data) in the corresponding service. 
The controller should be as thin as possible and only used to declare the API usage and transform request data to the service layer. This will help to make the code cleaner and more maintainable in the future.

joannechen1223 commented on apps/recnet-api/src/modules/article/article.service.ts:
```diff
@@ -64,4 +94,35 @@ export class ArticleService {
+  ) {
+    const existing = await this.articleRepository.findArticleByLink(link);
+    if (!existing) {
+      throw new NotFoundException(`Article not found by link=${link}`);
```
In our backend system, we wrapped a customized error, `RecnetError` which will be handled in the shared exception filter.

joannechen1223 reviewed (commented):
Thanks for your hard work! It was a great initiative. I left some comments related to the `recnet-api` design and development, summarized to two major perspectives here from my point of view.
1. For getting article by link, if the requirement is NOT to integrate the digital library and only query from DB, I suggest having a separate API and method to handle it. (Feel free to change to the original method name if you have a better naming idea.)
2. For updating article API, I suggest updating by article ID instead of the link. Some edge cases were provided in the comments.

If you have any questions, please let me know.

mistydqwq commented:
@swh00tw @joannechen1223 Thanks a lot for your review! I've already started to work on your comments, and fixed some problems locally. I will resolve the remaining problems these days. Btw I left some comments in the useDigitalLibrary or create a new api part, which might need to be discussed.

mistydqwq commented:
> I haven't started code review yet, just run on my machine. The UI looks great!! There are some small nitpicks
> 
> 1. For the "month" field, although we store 0-indexed number in DB, it's better to display 1-indexed number in UI (you can apply the offset behind the scene) for the sake of UX. (If you can use the same Dropdown component in `NewArticleForm.tsx` that would be even better)
> 
> <img alt="Screenshot 2025-02-04 at 5 12 55 PM" width="754" src="https://private-user-images.githubusercontent.com/71842426/409753093-fa848655-1bdd-4d40-bd51-fa2d9800d786.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyMTEzNjYsIm5iZiI6MTczOTIxMTA2NiwicGF0aCI6Ii83MTg0MjQyNi80MDk3NTMwOTMtZmE4NDg2NTUtMWJkZC00ZDQwLWJkNTEtZmEyZDk4MDBkNzg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEwVDE4MTEwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA3ZmU5MDg4ZjQxN2EyYTJiMzk5ZTJhOThjYzVkZTA2M2Y3NjI2NTIzMjNiODg0NjQxYmQyN2FjYTQ3YWEzNzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.m7K-d8dtWmBH2UbmVDthDF5LnwxQj9lE-sxTyGnKzP8"> <img alt="Screenshot 2025-02-04 at 5 16 15 PM" width="280" src="https://private-user-images.githubusercontent.com/71842426/409753940-fcc5fbbc-8663-4663-bfe1-56490d02e447.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyMTEzNjYsIm5iZiI6MTczOTIxMTA2NiwicGF0aCI6Ii83MTg0MjQyNi80MDk3NTM5NDAtZmNjNWZiYmMtODY2My00NjYzLWJmZTEtNTY0OTBkMDJlNDQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEwVDE4MTEwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI2YTQzM2NkN2RjMzI5MWNjY2U4Y2VkOWM0YTVlMzA4Y2QwZmFmNjJmZTEyN2RhOGVkMzE4MjlkNzE5OTkyMTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.fcKyeYqXYPhw6lWFsua9dZcuV4P8UHthiae7mXL5nhA">
> 2. After I click "Save changes", the form disappears. Is it intended behavior? I felt like the disappearing behavior is weird and leaving the form there should be better. @joannechen1223 any thoughts?

Good idea, I've changed the month field to dropdown list. And for the second question, I actually intended to do it. Because this feature I think is mainly used to modify the article's entry. At first I just think if we changed the article's link, and the form doesn't disappear, the form's link will not be the same as the link in the search box, which could also be a little bit weird. But If you think it's better to keep the form after saving the changes, I can do it.

swh00tw commented:
> > I haven't started code review yet, just run on my machine. The UI looks great!! There are some small nitpicks
> > 
> > 1. For the "month" field, although we store 0-indexed number in DB, it's better to display 1-indexed number in UI (you can apply the offset behind the scene) for the sake of UX. (If you can use the same Dropdown component in `NewArticleForm.tsx` that would be even better)
> > 
> > <img alt="Screenshot 2025-02-04 at 5 12 55 PM" width="754" src="https://private-user-images.githubusercontent.com/71842426/409753093-fa848655-1bdd-4d40-bd51-fa2d9800d786.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyMTEzNjYsIm5iZiI6MTczOTIxMTA2NiwicGF0aCI6Ii83MTg0MjQyNi80MDk3NTMwOTMtZmE4NDg2NTUtMWJkZC00ZDQwLWJkNTEtZmEyZDk4MDBkNzg2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEwVDE4MTEwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA3ZmU5MDg4ZjQxN2EyYTJiMzk5ZTJhOThjYzVkZTA2M2Y3NjI2NTIzMjNiODg0NjQxYmQyN2FjYTQ3YWEzNzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.m7K-d8dtWmBH2UbmVDthDF5LnwxQj9lE-sxTyGnKzP8"> <img alt="Screenshot 2025-02-04 at 5 16 15 PM" width="280" src="https://private-user-images.githubusercontent.com/71842426/409753940-fcc5fbbc-8663-4663-bfe1-56490d02e447.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkyMTEzNjYsIm5iZiI6MTczOTIxMTA2NiwicGF0aCI6Ii83MTg0MjQyNi80MDk3NTM5NDAtZmNjNWZiYmMtODY2My00NjYzLWJmZTEtNTY0OTBkMDJlNDQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjEwVDE4MTEwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI2YTQzM2NkN2RjMzI5MWNjY2U4Y2VkOWM0YTVlMzA4Y2QwZmFmNjJmZTEyN2RhOGVkMzE4MjlkNzE5OTkyMTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.fcKyeYqXYPhw6lWFsua9dZcuV4P8UHthiae7mXL5nhA">
> > 2. After I click "Save changes", the form disappears. Is it intended behavior? I felt like the disappearing behavior is weird and leaving the form there should be better. @joannechen1223 any thoughts?
> 
> Good idea, I've changed the month field to dropdown list. And for the second question, I actually intended to do it. Because this feature I think is mainly used to modify the article's entry. At first I just think if we changed the article's link, and the form doesn't disappear, the form's link will not be the same as the link in the search box, which could also be a little bit weird. But If you think it's better to keep the form after saving the changes, I can do it.

GJ!

For the second point, yeah I think it's a bit tricky. I would say leave them there for now. Don't do any changes until you figure out how to handle this kind of edge case. It's not super critical since we can still use this feature.

swh00tw commented:
Btw, can you push your changes to this branch? I saw you resolved a few comments but we can't see your updated code.

mistydqwq commented:
> Btw, can you push your changes to this branch? I saw you resolved a few comments but we can't see your updated code.

For sure, I've already pushed my current code. But somehow got the Vercel Issue. I tried to view details but got 404. It works good in my environment. May need some help about this question.

swh00tw commented:
> > Btw, can you push your changes to this branch? I saw you resolved a few comments but we can't see your updated code.
> 
> For sure, I've already pushed my current code. But somehow got the Vercel Issue. I tried to view details but got 404. It works good in my environment. May need some help about this question.

Hi, I saw the amount of lines of changes in `pnpm-lock.yaml` looks not normal. So, I tried delete it on my machine, then run `pnpm install` and produce a new `pnpm-lock.yaml` and push to another branch for testing. Then, the build process succeed. 

The branch is here: https://github.com/lil-lab/recnet/tree/feat/admin-manage-article-test

Therefore, I think there is something to do with your local pnpm version. Can you report your pnpm version? You can run `pnpm -v` to know.

swh00tw reviewed (changes requested):
Great! Some small things need to be finished
1. clear unused comment
2. Whether or not use param `fetchFromDigitalLibrary` in `article.service.ts` (I've shared my pov above)

mistydqwq commented:
> > > Btw, can you push your changes to this branch? I saw you resolved a few comments but we can't see your updated code.
> > 
> > 
> > For sure, I've already pushed my current code. But somehow got the Vercel Issue. I tried to view details but got 404. It works good in my environment. May need some help about this question.
> 
> Hi, I saw the amount of lines of changes in `pnpm-lock.yaml` looks not normal. So, I tried delete it on my machine, then run `pnpm install` and produce a new `pnpm-lock.yaml` and push to another branch for testing. Then, the build process succeed.
> 
> The branch is here: https://github.com/lil-lab/recnet/tree/feat/admin-manage-article-test
> 
> Therefore, I think there is something to do with your local pnpm version. Can you report your pnpm version? You can run `pnpm -v` to know.

Yeah, my pnpm version is 8.15.2, which version should I use?

swh00tw commented:
> > > > Btw, can you push your changes to this branch? I saw you resolved a few comments but we can't see your updated code.
> > > 
> > > 
> > > For sure, I've already pushed my current code. But somehow got the Vercel Issue. I tried to view details but got 404. It works good in my environment. May need some help about this question.
> > 
> > 
> > Hi, I saw the amount of lines of changes in `pnpm-lock.yaml` looks not normal. So, I tried delete it on my machine, then run `pnpm install` and produce a new `pnpm-lock.yaml` and push to another branch for testing. Then, the build process succeed.
> > The branch is here: https://github.com/lil-lab/recnet/tree/feat/admin-manage-article-test
> > Therefore, I think there is something to do with your local pnpm version. Can you report your pnpm version? You can run `pnpm -v` to know.
> 
> Yeah, my pnpm version is 8.15.2, which version should I use?

Hmm weird, it should work. Can you delete `pnpm-lock.yaml` and run `pnpm install` and push code again

mistydqwq commented:
> > > > > Btw, can you push your changes to this branch? I saw you resolved a few comments but we can't see your updated code.
> > > > 
> > > > 
> > > > For sure, I've already pushed my current code. But somehow got the Vercel Issue. I tried to view details but got 404. It works good in my environment. May need some help about this question.
> > > 
> > > 
> > > Hi, I saw the amount of lines of changes in `pnpm-lock.yaml` looks not normal. So, I tried delete it on my machine, then run `pnpm install` and produce a new `pnpm-lock.yaml` and push to another branch for testing. Then, the build process succeed.
> > > The branch is here: https://github.com/lil-lab/recnet/tree/feat/admin-manage-article-test
> > > Therefore, I think there is something to do with your local pnpm version. Can you report your pnpm version? You can run `pnpm -v` to know.
> > 
> > 
> > Yeah, my pnpm version is 8.15.2, which version should I use?
> 
> Hmm weird, it should work. Can you delete `pnpm-lock.yaml` and run `pnpm install` and push code again

I tried twice, but it didn't work. No idea what's the problem. Or maybe you can check the current changes and change the pnpm-lock.yaml in your environment, help me commit the change to this PR and we can finish this PR.

swh00tw commented on apps/recnet/src/server/routers/article.ts:
```diff
@@ -12,13 +13,24 @@ export const articleRouter = router({
       const { data } = await recnetApi.get(`/articles`, {
         params: {
           link,
+          useDigitalLibrary: useDigitalLibraryFallback,
```
```suggestion
          useDigitalLibraryFallback,
```
If the param name has been changed to `useDigitalLibraryFallback`, why do we still have `useDigitalLibrary` here?

swh00tw replied:
I've tested your code by running applications locally, this caused bug. Since the backend expect the param name be `useDigitalLibraryFallback` and you pass `useDigitalLibrary`, backend won't be able to receive the boolean value you want to pass.

I added the code here to test
![Screenshot 2025-02-13 at 9 16 08 PM](https://github.com/user-attachments/assets/544a9149-450e-47d1-826b-bc121236cfc4)
The log from `recnet-api`
![Screenshot 2025-02-13 at 9 16 47 PM](https://github.com/user-attachments/assets/faf19c2e-3ebd-41f9-8fb2-9a2ddbd6af9e)
Please test your code and make sure everything works correctly.

swh00tw commented on apps/recnet-api/src/modules/article/article.service.ts:
```diff
@@ -64,4 +65,39 @@ export class ArticleService {
       doi: null,
     };
   }
+
+  // public async getDbArticleByLink(
+  //   link: string
+  // ): Promise<GetArticleByLinkResponse> {
+  //   const article = await this.articleRepository.findArticleByLink(link);
+  //   if (article) {
+  //     return { article };
+  //   }
+  //
+  //   // If no article found and no metadata available, return null
+  //   return {
+  //     article: null,
+  //   };
+  // }
+
```
Please remove unused code.

swh00tw commented:
> > > > > > Btw, can you push your changes to this branch? I saw you resolved a few comments but we can't see your updated code.
> > > > > 
> > > > > 
> > > > > For sure, I've already pushed my current code. But somehow got the Vercel Issue. I tried to view details but got 404. It works good in my environment. May need some help about this question.
> > > > 
> > > > 
> > > > Hi, I saw the amount of lines of changes in `pnpm-lock.yaml` looks not normal. So, I tried delete it on my machine, then run `pnpm install` and produce a new `pnpm-lock.yaml` and push to another branch for testing. Then, the build process succeed.
> > > > The branch is here: https://github.com/lil-lab/recnet/tree/feat/admin-manage-article-test
> > > > Therefore, I think there is something to do with your local pnpm version. Can you report your pnpm version? You can run `pnpm -v` to know.
> > > 
> > > 
> > > Yeah, my pnpm version is 8.15.2, which version should I use?
> > 
> > 
> > Hmm weird, it should work. Can you delete `pnpm-lock.yaml` and run `pnpm install` and push code again
> 
> I tried twice, but it didn't work. No idea what's the problem. Or maybe you can check the current changes and change the pnpm-lock.yaml in your environment, help me commit the change to this PR and we can finish this PR.

I can do that but this is not the best solution, if you cannot find the root cause, you are gonna encounter this over and over again. Have you tried search for solutions? Here are somethings I found, maybe it can help:
- https://github.com/npm/cli/issues/6785
- https://github.com/npm/cli/issues/6785#issuecomment-1937801989
- https://github.com/npm/cli/issues/6785#issuecomment-1720293938

Please fix the code first and if you still can't solve it tomorrow, I can help you commit the code. But still, you need to solve this next time by yourself.

mistydqwq commented:
> > > > > > > Btw, can you push your changes to this branch? I saw you resolved a few comments but we can't see your updated code.
> > > > > > 
> > > > > > 
> > > > > > For sure, I've already pushed my current code. But somehow got the Vercel Issue. I tried to view details but got 404. It works good in my environment. May need some help about this question.
> > > > > 
> > > > > 
> > > > > Hi, I saw the amount of lines of changes in `pnpm-lock.yaml` looks not normal. So, I tried delete it on my machine, then run `pnpm install` and produce a new `pnpm-lock.yaml` and push to another branch for testing. Then, the build process succeed.
> > > > > The branch is here: https://github.com/lil-lab/recnet/tree/feat/admin-manage-article-test
> > > > > Therefore, I think there is something to do with your local pnpm version. Can you report your pnpm version? You can run `pnpm -v` to know.
> > > > 
> > > > 
> > > > Yeah, my pnpm version is 8.15.2, which version should I use?
> > > 
> > > 
> > > Hmm weird, it should work. Can you delete `pnpm-lock.yaml` and run `pnpm install` and push code again
> > 
> > 
> > I tried twice, but it didn't work. No idea what's the problem. Or maybe you can check the current changes and change the pnpm-lock.yaml in your environment, help me commit the change to this PR and we can finish this PR.
> 
> I can do that but this is not the best solution, if you cannot find the root cause, you are gonna encounter this over and over again. Have you tried search for solutions? Here are somethings I found, maybe it can help:
> 
> * [[BUG] all commands fail: `TypeError: LRU is not a constructor` npm/cli#6785](https://github.com/npm/cli/issues/6785)
> * [[BUG] all commands fail: `TypeError: LRU is not a constructor` npm/cli#6785 (comment)](https://github.com/npm/cli/issues/6785#issuecomment-1937801989)
> * [[BUG] all commands fail: `TypeError: LRU is not a constructor` npm/cli#6785 (comment)](https://github.com/npm/cli/issues/6785#issuecomment-1720293938)
> 
> Please fix the code first and if you still can't solve it tomorrow, I can help you commit the code. But still, you need to solve this next time by yourself.

I cannot see the error information in vercel, maybe I don't have the permission. So if TypeError: LRU is not a constructor is the problem, the root I think is the two dependencies below in the pnpm-lock.yaml have incompatible lru-cache version. If this is the cause, then the problem could be either pnpm/node version problem or package.json problem. But I have never changed the package.json, and you said my pnpm version should be fine, maybe it's node or something I don't know. I think I will try to download the pnpm-lock.yaml from main branch and run pnpm install --frozen-lockfile everytime.

  /@babel/helper-compilation-targets@7.26.5:
    resolution: {integrity: sha512-IXuyn5EkouFJscIDuFF5EsiSolseme1s0CZB+QxVugqJLYmKdxI1VfIBOst0SUu4rnk2Z7kqTwmoO1lp3HIfnA==}
    engines: {node: '>=6.9.0'}
    dependencies:
      '@babel/compat-data': 7.26.8
      '@babel/helper-validator-option': 7.25.9
      browserslist: 4.24.4
      lru-cache: 5.1.1
      semver: 6.3.1

  /@asamuzakjp/css-color@2.8.3:
    resolution: {integrity: sha512-GIc76d9UI1hCvOATjZPyHFmE5qhRccp3/zGfMPapK3jBi+yocEzp6BBB0UnfRYP9NP4FANqUZYb0hnfs3TM3hw==}
    dependencies:
      '@csstools/css-calc': 2.1.1(@csstools/css-parser-algorithms@3.0.4)(@csstools/css-tokenizer@3.0.3)
      '@csstools/css-color-parser': 3.0.7(@csstools/css-parser-algorithms@3.0.4)(@csstools/css-tokenizer@3.0.3)
      '@csstools/css-parser-algorithms': 3.0.4(@csstools/css-tokenizer@3.0.3)
      '@csstools/css-tokenizer': 3.0.3
      lru-cache: 10.4.3
    dev: true

swh00tw commented:
> > > > > > > > Btw, can you push your changes to this branch? I saw you resolved a few comments but we can't see your updated code.
> > > > > > > 
> > > > > > > 
> > > > > > > For sure, I've already pushed my current code. But somehow got the Vercel Issue. I tried to view details but got 404. It works good in my environment. May need some help about this question.
> > > > > > 
> > > > > > 
> > > > > > Hi, I saw the amount of lines of changes in `pnpm-lock.yaml` looks not normal. So, I tried delete it on my machine, then run `pnpm install` and produce a new `pnpm-lock.yaml` and push to another branch for testing. Then, the build process succeed.
> > > > > > The branch is here: https://github.com/lil-lab/recnet/tree/feat/admin-manage-article-test
> > > > > > Therefore, I think there is something to do with your local pnpm version. Can you report your pnpm version? You can run `pnpm -v` to know.
> > > > > 
> > > > > 
> > > > > Yeah, my pnpm version is 8.15.2, which version should I use?
> > > > 
> > > > 
> > > > Hmm weird, it should work. Can you delete `pnpm-lock.yaml` and run `pnpm install` and push code again
> > > 
> > > 
> > > I tried twice, but it didn't work. No idea what's the problem. Or maybe you can check the current changes and change the pnpm-lock.yaml in your environment, help me commit the change to this PR and we can finish this PR.
> > 
> > 
> > I can do that but this is not the best solution, if you cannot find the root cause, you are gonna encounter this over and over again. Have you tried search for solutions? Here are somethings I found, maybe it can help:
> > 
> > * [[BUG] all commands fail: `TypeError: LRU is not a constructor` npm/cli#6785](https://github.com/npm/cli/issues/6785)
> > * [[BUG] all commands fail: `TypeError: LRU is not a constructor` npm/cli#6785 (comment)](https://github.com/npm/cli/issues/6785#issuecomment-1937801989)
> > * [[BUG] all commands fail: `TypeError: LRU is not a constructor` npm/cli#6785 (comment)](https://github.com/npm/cli/issues/6785#issuecomment-1720293938)
> > 
> > Please fix the code first and if you still can't solve it tomorrow, I can help you commit the code. But still, you need to solve this next time by yourself.
> 
> I cannot see the error information in vercel, maybe I don't have the permission. So if TypeError: LRU is not a constructor is the problem, the root I think is the two dependencies below in the pnpm-lock.yaml have incompatible lru-cache version. If this is the cause, then the problem could be either pnpm/node version problem or package.json problem. But I have never changed the package.json, and you said my pnpm version should be fine, maybe it's node or something I don't know. I think I will try to download the pnpm-lock.yaml from main branch and run pnpm install --frozen-lockfile everytime.
> 
> /@babel/helper-compilation-targets@7.26.5: resolution: {integrity: sha512-IXuyn5EkouFJscIDuFF5EsiSolseme1s0CZB+QxVugqJLYmKdxI1VfIBOst0SUu4rnk2Z7kqTwmoO1lp3HIfnA==} engines: {node: '>=6.9.0'} dependencies: '@babel/compat-data': 7.26.8 '@babel/helper-validator-option': 7.25.9 browserslist: 4.24.4 lru-cache: 5.1.1 semver: 6.3.1
> 
> /@asamuzakjp/css-color@2.8.3: resolution: {integrity: sha512-GIc76d9UI1hCvOATjZPyHFmE5qhRccp3/zGfMPapK3jBi+yocEzp6BBB0UnfRYP9NP4FANqUZYb0hnfs3TM3hw==} dependencies: '@csstools/css-calc': 2.1.1(@csstools/css-parser-algorithms@3.0.4)(@csstools/css-tokenizer@3.0.3) '@csstools/css-color-parser': 3.0.7(@csstools/css-parser-algorithms@3.0.4)(@csstools/css-tokenizer@3.0.3) '@csstools/css-parser-algorithms': 3.0.4(@csstools/css-tokenizer@3.0.3) '@csstools/css-tokenizer': 3.0.3 lru-cache: 10.4.3 dev: true

Yeah, you can try it. I saw someone just reinstall node, npm, and pnpm. Might be worth to try it.

mistydqwq commented:
Seems like we can finish this PR?
